### PR TITLE
bunny is in c:

### PR DESCRIPTION
**Pull request by ItsRaijin**
yipppppppeeeeeeeeeeeeeeeeeeee

### CODE REVIEW

File: src/controller.ts [hunk 1]
```diff
@@ -1,15 +1,16 @@
...
 export class Controller implements MinigameDelegate {
   private currentMinigame?: Minigame;
 
   constructor(private readonly app: Application) { }
 
-  start() {
-    this.currentMinigame = new KeyboardMinigame(this.app, this);
+  async start() {
+    this.currentMinigame = new DragMinigame(this.app, this);
+    await this.currentMinigame.preLoad();
```
Comment (zainafzal08) [inline src/controller.ts:13]: nitpick: Remove capital L and just use preload (all lower case)

File: src/minigames/dragMinigame.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
+import { Minigame } from "./minigame";
+import { Application, Assets, FederatedMouseEvent, Graphics, Sprite, Texture } from "pixi.js";
+export class DragMinigame extends Minigame{
+
```
Comment (zainafzal08) [inline src/minigames/dragMinigame.ts:4]: nitpick: remove new line

File: src/minigames/dragMinigame.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    override async preLoad(): Promise<void> {
+        this.texture = await Assets.load('https://pixijs.com/assets/bunny.png');
+        
+    }
+    
+    override attach() {
+        super.attach();
+        this.app.stage.eventMode = 'static';
+        this.app.stage.hitArea = this.app.screen;
+        this.app.stage.on('pointerup', ()=> void this.onDragEnd());
```
Comment (zainafzal08) [inline src/minigames/dragMinigame.ts:18]: nitpick: space after the `()` so it's `() => ...`

File: src/minigames/dragMinigame.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        this.app.stage.on('pointerupoutside', ()=> void this.onDragEnd());
+    }
+
+    protected override populateContainer(): void {
+        const square = new Sprite(this.texture);
+        const x = (this.app.screen.width - 100) * Math.random();
+        const y = (this.app.screen.height - 100) * Math.random();
+        square.anchor.set(0.5);
+        //square.rect(x, y, 100, 100);
```
Comment (zainafzal08) [inline src/minigames/dragMinigame.ts:27]: Remove commented lines

File: src/minigames/dragMinigame.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    override attach() {
+        super.attach();
+        this.app.stage.eventMode = 'static';
+        this.app.stage.hitArea = this.app.screen;
+        this.app.stage.on('pointerup', ()=> void this.onDragEnd());
+        this.app.stage.on('pointerupoutside', ()=> void this.onDragEnd());
+    }
+
+    protected override populateContainer(): void {
+        const square = new Sprite(this.texture);
```
Comment (zainafzal08) [inline src/minigames/dragMinigame.ts:23]: nitpick: rename to bunny

File: src/minigames/dragMinigame.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        //square.fill(0xde3249);
+        square.x = x;
+        square.y = y;
+        square.scale.set(10);
+        
+        square.eventMode = 'static';
+        square.on('pointerdown', () => { 
+            this.dragObject = square;
+            this.app.stage.on("pointermove", this.dragListener);
+            console.log(this.dragObject);
```
Comment (zainafzal08) [inline src/minigames/dragMinigame.ts:37]: Nitpick: do we need this console.log?

File: src/minigames/dragMinigame.ts [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        });
+        this.container.addChild(square);
+    }
+    //private dragObject: Graphics;
```
Comment (zainafzal08) [inline src/minigames/dragMinigame.ts:41]: remove

File: src/minigames/dragMinigame.ts [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
```
Comment (zainafzal08) [inline src/minigames/dragMinigame.ts:42]: 1 new line between methods

File: src/minigames/dragMinigame.ts [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+            this.app.stage.on("pointermove", this.dragListener);
+            console.log(this.dragObject);
+        });
+        this.container.addChild(square);
+    }
+    //private dragObject: Graphics;
+    
+
+    private onDragEnd() {
+        if (this.dragObject)
```
Comment (zainafzal08) [inline src/minigames/dragMinigame.ts:45]: nitpick: use the The One True Brace Style ™

```
if (this.dragObject) {
   ...
}
```

File: src/minigames/dragMinigame.ts [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+        {
+            console.log("ENDED");
+            this.app.stage.off('pointermove', this.dragListener);
+            this.dragObject = null;
+        }
+    }
+
+    private onDrag(event: FederatedMouseEvent) {
+        console.log(this.dragObject);
```
Comment (zainafzal08) [inline src/minigames/dragMinigame.ts:54]: nitpick: do we need this? Spams the console maybe better to leave it.

File: src/minigames/minigame.ts [hunk 11]
```diff
@@ -12,6 +12,10 @@ export class Minigame {
     this.container = new Container();
   }
 
+  async preLoad() {
+
```
Comment (zainafzal08) [inline src/minigames/minigame.ts:16]: nit: Add a comment saying "load assets for a minigame here"

File: src/minigames/dragMinigame.ts [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+import { Application, Assets, FederatedMouseEvent, Graphics, Sprite, Texture } from "pixi.js";
+export class DragMinigame extends Minigame{
+
+    protected dragObject: Sprite | null = null;
+    private texture: Texture | null = null;
+    private dragListener = (e: FederatedMouseEvent) => void this.onDrag(e);
+
+    override async preLoad(): Promise<void> {
+        this.texture = await Assets.load('https://pixijs.com/assets/bunny.png');
+
```
Comment (zainafzal08) [inline src/minigames/dragMinigame.ts:11]: nitpick: remove new line

**Review by zainafzal08 — APPROVED** [review]

File: src/index.ts [hunk 13]
```diff
@@ -1,4 +1,4 @@
-import { Application } from 'pixi.js';
+import { Application, Assets } from 'pixi.js';
```
Comment (josephkimsungha) [inline src/index.ts:1]: pls undo

File: src/minigames/dragMinigame.ts [hunk 14]
```diff
@@ -0,0 +1,59 @@
+import { Minigame } from "./minigame";
+import { Application, Assets, FederatedMouseEvent, Graphics, Sprite, Texture } from "pixi.js";
+export class DragMinigame extends Minigame{
```
Comment (josephkimsungha) [inline src/minigames/dragMinigame.ts:3]: nit: newline above and space after `Minigame`

File: src/minigames/dragMinigame.ts [hunk 15]
```diff
@@ -0,0 +1,59 @@
+import { Minigame } from "./minigame";
+import { Application, Assets, FederatedMouseEvent, Graphics, Sprite, Texture } from "pixi.js";
```
Comment (josephkimsungha) [inline src/minigames/dragMinigame.ts:2]: nit: I think some of these are not being used, e.g. Application, Graphics. Can we remove them?

File: src/controller.ts [hunk 16]
```diff
@@ -1,15 +1,16 @@
...
 /** Controls the flow of the game. */
 export class Controller implements MinigameDelegate {
   private currentMinigame?: Minigame;
 
   constructor(private readonly app: Application) { }
 
-  start() {
-    this.currentMinigame = new KeyboardMinigame(this.app, this);
+  async start() {
+    this.currentMinigame = new DragMinigame(this.app, this);
```
Comment (josephkimsungha) [inline src/controller.ts:12]: After you merge/rebase in changes from #12 (if it happens to land before yours), you'll want to add DragMinigame to the list of ALL_MINIGAMES and `await preload` in the `populateMinigameQueue` method, I reckon. If this lands before #12 no action required

File: src/minigames/dragMinigame.ts (hunk 6 above)
Comment (josephkimsungha) [reply src/minigames/dragMinigame.ts:37]: I vote remove

File: src/minigames/dragMinigame.ts [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+    //private dragObject: Graphics;
+    
+
+    private onDragEnd() {
+        if (this.dragObject)
+        {
+            console.log("ENDED");
+            this.app.stage.off('pointermove', this.dragListener);
+            this.dragObject = null;
+        }
```
Comment (josephkimsungha) [inline src/minigames/dragMinigame.ts:50]: It's a bit neater if we flip the conditional and return early if there's no dragObject.

```suggestion
        if (!this.dragObject) return;
        
        this.app.stage.off('pointermove', this.dragListener);
        this.dragObject = null;
```

File: src/minigames/dragMinigame.ts (hunk 10 above)
Comment (josephkimsungha) [reply src/minigames/dragMinigame.ts:54]: yes let's remove

File: src/minigames/dragMinigame.ts [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    private onDrag(event: FederatedMouseEvent) {
+        console.log(this.dragObject);
+        if (this.dragObject) {
+            this.dragObject.parent.toLocal(event.global, null, this.dragObject.position);
+        }
+    }
+}
```
Comment (josephkimsungha) [inline src/minigames/dragMinigame.ts:59]: nit: newline at eof

**Review by josephkimsungha — APPROVED** [review]

File: src/controller.ts (hunk 16 above)
Comment (josephkimsungha) [reply src/controller.ts:12]: Actually I just realised this minigame has no win or lose condition so it locks players in lol. Maybe we should leave it out of ALL_MINIGAMES to begin with

**Review by josephkimsungha — APPROVED** [review]

File: src/minigames/dragMinigame.ts [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+        //square.rect(x, y, 100, 100);
+        //square.fill(0xde3249);
+        square.x = x;
+        square.y = y;
+        square.scale.set(10);
+        
+        square.eventMode = 'static';
+        square.on('pointerdown', () => { 
+            this.dragObject = square;
+            this.app.stage.on("pointermove", this.dragListener);
```
Comment (josephkimsungha) [inline src/minigames/dragMinigame.ts:36]: It probably isn't a good idea to add these listeners to `this.app.stage`. Can we put them on `this.container` instead?

**Pull request closed without merging**